app/models/purpose.rb

```diff
@@ -59,7 +59,11 @@ def source_plate(labware)
     ActiveSupport::Deprecation.silence do
       # Rails 6 lets us do this:
       # ActiveSupport::Deprecation.allow(:stock_plate) do
-      source_purpose_id.present? ? labware.ancestor_of_purpose(source_purpose_id) : labware.stock_plate
+      if source_purpose_id.present?
+        labware.ancestors_of_purpose(source_purpose_id)
+      elsif labware.respond_to?(:stock_plate)
+        labware.stock_plate
```

Comment:
This can simply become labware.try(:stock_plate) (Similarly below)

```
[1] pry(main)> TubeRack.new.try(:stock_plate)
   (1.7ms)  SET NAMES utf8,  @@SESSION.sql_mode = 'TRADITIONAL',  @@SESSION.group_concat_max_len = 67108864, @@SESSION.sql_auto_is_null = 0, @@SESSION.wait_timeout = 2147483
=> nil
```

Although I'm wondering if we should be defining stock_plate on labware (Although the fact we've deprecated it makes me wary)

Reply:
I was also going to suggest that we might be able to work out the source plates, but from glancing at the way the features actually used, it wouldn't be much use.

I think we can actually avoid needing to use this method at all here, and instead migrate to ancestors_of_purpose. However I need to work out some of the details.

Background: This is mostly used to generate links to plate summaries, which provide an overview of all actions that happened to a plate as it was processed in a pipeline. As a result, each plate or tube tries to link back to the plate at the top of the pipeline, using the source purpose. However this is a bit hard coded, and doesn't really take into account all the variation that can happen now. Instead it would be better to grab a list of any ancestor that could theoretically be of interest, based on a list of all suitable plates.

`labware.ancestor_of_purpose(PlatePurpose.considered_stock_plate)` would get us most the way there, but there might be a few edge cases I'm not thinking of at the moment.

Reply:
This sounds like a lot of change to resolve an issue that hasn't been reported yet.  I think given that:
a) `:stock_plate` has been deprecated
b) The TubeRack is able to load now with just the use of `:try`
that we should go for the simplest solution for now.  If it's desirable to see the source for TubeRacks and this doesn't cover that use case, then that should be a different ticket (i.e. not this bug report ticket)